opentracing-apache-httpclient/src/main/java/io/opentracing/contrib/apache/TracingClientExec.java

```diff
@@ -0,0 +1,151 @@
+   * Tracing {@link ClientExecChain} is executed after redirect exec, so on redirects it is called
+   * multiple times. This is used as an id for {@link HttpClientContext#setAttribute(String, Object)}
+   * to store number of redirects.
+   */
+  protected static final String REDIRECT_COUNT = TracingHttpClientBuilder.class.getName() + ".redirectCount";
+
+  private final RedirectStrategy redirectStrategy;
+  private final ClientExecChain requestExecutor;
+  private final boolean redirectHandlingDisabled;
+
+  private final Tracer tracer = GlobalTracer.get();
+  private final SpanManager spanManager = DefaultSpanManager.getInstance();
+  private final List<ApacheClientSpanDecorator> spanDecorators;
+
+  public TracingClientExec(ClientExecChain clientExecChain, RedirectStrategy redirectStrategy,
+      boolean redirectHandlingDisabled, List<ApacheClientSpanDecorator> spanDecorators) {
+    this.requestExecutor = clientExecChain;
+    this.redirectStrategy = redirectStrategy;
+    this.redirectHandlingDisabled = redirectHandlingDisabled;
+    this.spanDecorators = new ArrayList<>(spanDecorators);
+  }
+
+  @Override
+  public CloseableHttpResponse execute(HttpRoute route,
+      HttpRequestWrapper request,
+      HttpClientContext clientContext,
+      HttpExecutionAware execAware) throws IOException, HttpException {
+
+    Span localSpan = clientContext.getAttribute(ACTIVE_SPAN, Span.class);
+    CloseableHttpResponse response = null;
+    try {
+      if (localSpan != null) {
+        return (response = handleNetworkProcessing(localSpan, route, request, clientContext, execAware));
+      }
+
+      localSpan = handleLocalSpan(request, clientContext);
+      return (response = handleNetworkProcessing(localSpan, route, request, clientContext, execAware));
+    } catch (Exception e) {
+      localSpan.finish();
+      throw e;
+    } finally {
+      if (response != null) {
+        /**
+         * This exec runs after {@link org.apache.http.impl.execchain.RedirectExec} which loops
+         * until there is no redirect or reaches max redirect count.
+         * {@link RedirectStrategy} is used to decide whether localSpan should be finished or not.
+         * If there is a redirect localSpan is not finished and redirect is logged.
+         */
+        Integer redirectCount = clientContext.getAttribute(REDIRECT_COUNT, Integer.class);
+        if (!redirectHandlingDisabled &&
+            clientContext.getRequestConfig().isRedirectsEnabled() &&
+            redirectStrategy.isRedirected(request, response, clientContext) &&
+            ++redirectCount < clientContext.getRequestConfig().getMaxRedirects()) {
+
+          clientContext.setAttribute(REDIRECT_COUNT, redirectCount);
+        } else {
+          localSpan.finish();
+        }
+      }
+    }
+  }
+
+  protected Span handleLocalSpan(HttpRequest httpRequest, HttpClientContext clientContext) {
+    Tracer.SpanBuilder spanBuilder = tracer.buildSpan(httpRequest.getRequestLine().getMethod())
```

Comment:
Not sure about creating the local span - I can see a possible benefit if there is a redirect, otherwise it appears (correct me if wrong) that a non-redirected invocation will now result in two spans - one internal and one client span?

Another option would be to just identify reference from subsequent client spans to the span they are redirected from - which could be a reusable type that could also be used in fallback/retry situations.